Fix Stripe nudge autosave and redirects in FSE

stacimc opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->
* The existing redirects after Stripe connection only work in the block editor. In the site editor, the redirect always falls back to the `earn/payments` settings page. The templates also fail to auto-save before you are taken to the Stripe connection form.
* The Recurring Payments block in particular was also not redirecting correctly in the _block-editor_, and thus also fell back to the settings page.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Refactors the logic which modifies the Stripe connection URL to redirect correctly out into a shared function
    * Adds state to the redirect url, which will be used by the API to redirect back to the site editor. This implementation chosen to avoid disrupting how the redirects currently work for the block editor.
* Refactors autosave and redirection logic, introduced to upgrade nudges in #19636 to handle both site and block editor, out into a reusable hook
* Updates the shared Stripe nudge to use both of these new shared functions (this is the old-style nudge seen in the Donations block, see screenshots)
* Updates the `Connect Stripe` button in the toolbars for Premium Content and Recurring Payments to use these new shared functions


**Not included: API change to actually redirect correctly in the site editor. SEE D60831**

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

You will need a test site with a blocks theme, premium plan and without a Stripe connection.
* Between each test of the Stripe nudge, you will need to disconnect the Stripe account you've just added from the `earn/payments` settings page. You may need to refresh this page a few times for the option to disconnect to appear.
* Because this PR does not include the backend changes to redirect to the site-editor, you should only test the autosave feature, and that it does not break existing redirects. The Upgrade nudges need to be regression tested due to the refactor.
* Tests are done with Premium Content, Recurring Payments, and Donations because the refactor touches these three blocks differently.

**Test Stripe Nudges**

1. Within the site editor (`<blog_url>/wp-admin/admin.php?page=gutenberg-edit-site`), insert a Payments block. Make a few modifications to the page, including to other template parts. Note the changes you've made, but do not save.
2. Press the `Connect Stripe` button in the block toolbar. Note that the Save icon in the top right of the page shows that it is in a saving state.
3. Complete or skip (if sandboxed) the Stripe connection form. Verify you are redirected to the **`earn/payments` page for now.**
4. Verify that the changes you made prior to connecting were saved.
5. Repeat steps 1-4 with a Premium Content and Donations block. (For Donations, the button will be in a banner, rather than the block toolbar).
6. Repeat steps 1- 4 using at least one of the blocks within the Calypso iframe (ie visit the site-editor at `wordpress.com/site-editor/<blog_url>`)
7. Within the **block editor**, insert a Payments block. Make a few modifications to the post before using the block to Connect to Stripe. Verify you are redirected to the post in the block editor, and your changes were saved
8. Repeat test 6 with a Premium Content and Donations block.

**Test Upgrade Nudges**

Repeat all tests from #19636

stacimc committed:
[not verified] Refactor getCurrentUrl out into shared function

stacimc committed:
[not verified] Use getConnectUrl to redirect correctly from Payments block

stacimc committed:
[not verified] Update StripeNudge to use shared getConnectUrl

stacimc committed:
[not verified] Set state needed for redirect to site editor

stacimc committed:
[not verified] Add changelog

stacimc committed:
[not verified] Refactor autosave and redirect functionality into its own hook

stacimc committed:
[not verified] Update README

stacimc committed:
[not verified] Add autosave to Premium Content Stripe connect

stacimc committed:
[not verified] Add autosave to Payments Stripe connect

stacimc committed:
[not verified] Add autosave to shared Stripe nudge

ramonjd commented on projects/plugins/jetpack/extensions/blocks/premium-content/edit.js:
```diff
@@ -282,9 +287,19 @@ function Edit( props ) {
+		return false;
+	};
+
+	const [ , autosaveAndRedirect ] = useAutosaveAndRedirect( connectURL );
```
Is something missing here? One of `autosave | isRedirecting` or is the comma not required?

andrewserong replied:
I think it's for the array destructuring syntax since the hook returns an array instead of an object. Would it be worth updating the hook to return an object instead, since we aren't using array destructuring to rename the methods?

ramonjd replied:
Oh, the opening comma skips the first item. 👍 Thanks!
Carry on 👍

stacimc replied:
I'm skipping the first item (`autosave`) in the array destructuring with the placeholder comma. I'm actually unsure what the convention here should be -- `_autosave` instead to indicate it's unused for the linter?

stacimc replied:
+1 for the object return value!

ramonjd commented on projects/plugins/jetpack/extensions/blocks/premium-content/edit.js:
```diff
@@ -282,9 +287,19 @@ function Edit( props ) {
+		}
+
+		return false;
+	};
```
Not a biggie, but this could be simplified to:

```
const shouldShowConnectButton = () => 
    ! shouldUpgrade && apiState !== API_STATE_CONNECTED && connectURL
```

ramonjd commented on projects/plugins/jetpack/extensions/shared/use-autosave-and-redirect/index.js:
```diff
@@ -0,0 +1,91 @@
+	const [ isRedirecting, setIsRedirecting ] = useState( false );
+
+	const { isAutosaveablePost, isDirtyPost, currentPost } = useSelect( select => {
+		const editorSelector = select( 'core/editor' );
```
Another totally optional style change is to destructure the object:

```
const { getCurrentPost, isEditedPostAutosaveable, isEditedPostDirty } = select( 'core/editor' );
```

andrewserong replied:
@ramonjd I _think_ calling these methods within the useSelect is preferred (I remember seeing a similar example in p7H4VZ-2Q6-p2) so that changes in the value returned by the selector causes a re-render.

ramonjd replied:
That makes sense. Thanks for the context @andrewserong 👍

stacimc committed:
[not verified] Clean up conditional

andrewserong commented on projects/plugins/jetpack/extensions/shared/use-autosave-and-redirect/index.js:
```diff
@@ -0,0 +1,91 @@
+		} );
+	};
+
+	return [ autosave, autosaveAndRedirect, isRedirecting ];
```
Commented separately in https://github.com/Automattic/jetpack/pull/19671/files#r623519176, but would it make using the hook a bit neater if we returned an object with these methods instead of an array? I _think_ returning an array is more useful when the code consuming the hook intends to rename the methods (like when we use `useState`). For this hook, since we're keeping the method names the same, an object might be more straightforward so in consuming code we can use `const { autosaveAndRedirect } = useAutosaveAndRedirect( redirectUrl )` instead of needing to specify the position of the array to grab that method. (That said, I haven't written many custom hooks, so apologies if there's a reason you'd prefer to use an array!)

stacimc replied:
Not at all -- just a holdover from imitating the structure of the hook this was refactored out of. Returning an object looks much neater for the destructuring, thanks!

andrewserong commented on projects/plugins/jetpack/extensions/shared/currencies.js:
```diff
@@ -93,3 +98,40 @@ export function parseAmount( amount, currency ) {
+ * @param { string } connectURL - Stripe connect URL
+ * @returns { null | string } URL
+ */
+export function getConnectUrl( postId, connectURL ) {
```
Tiny nit: the other functions in this file are to do with dealing with currencies. Would it be worth moving this function to its own `get-connect-url.js` file?

andrewserong commented on projects/plugins/jetpack/extensions/shared/use-autosave-and-redirect/index.js:
```diff
@@ -0,0 +1,91 @@
+export default function useAutosaveAndRedirect( redirectUrl, onRedirect = noop ) {
+	const [ isRedirecting, setIsRedirecting ] = useState( false );
+
+	const { isAutosaveablePost, isDirtyPost, currentPost } = useSelect( select => {
```
Unfortunately it looks like using this hook is causing the Recurring Payments JS tests to fail. We might need to add some mocks to the existing tests?

![image](https://user-images.githubusercontent.com/14988353/116637537-858a5480-a9a7-11eb-98bc-7daa86e37a10.png)

ramonjd replied:
If it's of any use, I created a hook test in the [ GIF block](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/extensions/blocks/gif/test/use-fetch-giphy-data.js) for reference

andrewserong reviewed (changes requested):
Thanks for the detailed testing instructions again @stacimc! I like the refactor here, too, moving the save and redirect to a custom hook 👍.

Unfortunately it looks like the redirect to Stripe when clicking the Stripe Connect button doesn't work within the wordpress.com Calypso iframe (e.g. accessing the site via `wordpress.com/site-editor/<blog_url>`). It looks like attempting to access `window.top.location.href` throws a cross-origin frame error (in the UI this presents as the button not doing anything when I click it):

![image](https://user-images.githubusercontent.com/14988353/116639394-d00dd000-a9ab-11eb-9214-c3d6ab8e0a97.png)

I think we didn't run into this issue while testing the Upgrade nudges in the other PR, because for upgrade nudges, instead of doing the redirect from within the site editor iframe, it opens the checkout modal instead.

I'm not quite sure what the solution is, but perhaps there's a clue in the way the checkout modal works? E.g. can we fire off an action from the iframe to the parent frame/window to then perform the redirect? (I don't _quite_ understand how all these pieces connect, so I'm not sure if that's just another rabbit hole 🤔)

I also did a quick smoke test that the save behaviour is preserved for the upgrade nudges, too, so that looks like it's still working 👍

Happy to give all this another test on Monday!

ramonjd commented:
Just wanted to swing by and say that I tested this in conjunction with D60831-code

The site and block editor save/redirects work as expected for me (I tested the payments block only)

**Site Editor**
![Apr-30-2021 12-03-53](https://user-images.githubusercontent.com/6458278/116640070-51199700-a9ad-11eb-8318-a2fdb95d0f90.gif)

**Block Editor**
![Apr-30-2021 12-06-50](https://user-images.githubusercontent.com/6458278/116640117-70182900-a9ad-11eb-8546-2c37b686a4f9.gif)

ramonjd commented:
>  It looks like attempting to access window.top.location.href throws a cross-origin frame error

Maybe some of the new [iframe check methods](https://github.com/Automattic/jetpack/blob/011a67edbafbc809085b3e78847830a237d2101e/projects/plugins/jetpack/extensions/shared/block-editor-asset-loader.js#L54) will be helpful here too?

stacimc committed:
[not verified] Return autosave functions in an object for cleaner destructuring

stacimc committed:
[not verified] Move getConnectUrl to its own file

stacimc committed:
[not verified] Fix mocking of autosave in Payments tests

stacimc committed:
[not verified] Remove unauthorized check of parent window location

Looks like this check was originally used by the upgrade nudges to identify
when the block was in Calypso and prevent redirection. After adding in the
checkout modal it appears to be unnecessary (and upgrade specific).

This check was also causing CORS issues for Stripe nudges in the site
editor, due to the cross-origin iframe. Removing it resolves this bug.

andrewserong reviewed (approved):
Thanks for implementing all those changes @stacimc, this looks good to me, now! Tested with D60831-code applied.

✅ JS tests are passing
✅ Checked stripe connect, autosave + redirect is working correctly in `/wp-admin`
✅ Checked stripe connect, autosave + redirect is working correctly in `wordpress.com/site-editor/`
✅ Smoke tested the upgrade flow still works as expected

Great work on all this! 🎉

jeherve committed:
Merge remote-tracking branch 'origin/master' into fix/stripe-nudge-redirect-for-fse

jeherve committed:
Merge remote-tracking branch 'origin/master' into fix/stripe-nudge-redirect-for-fse

jeherve reviewed (approved):
Tests well for me. 🚢

stacimc committed:
Merge remote-tracking branch 'origin/master' into fix/stripe-nudge-redirect-for-fse

stacimc commented:
r226710-wpcom
